pkg/datapath/linux/probes/probes.go

```diff
@@ -386,10 +387,145 @@ func HaveBoundedLoops() error {
 }
 
 // HaveFibIfindex checks if kernel has d1c362e1dd68 ("bpf: Always return target
-// ifindex in bpf_fib_lookup") which is 5.10+. This got merged in the same kernel
-// as the new redirect helpers.
-func HaveFibIfindex() error {
-	return features.HaveProgramHelper(ebpf.SchedCLS, asm.FnRedirectPeer)
+// ifindex in bpf_fib_lookup") which is 5.10+ (and backports). It does so by
+// checking that ifindex has been updated as a result of the RIB lookup even
+// though the bpf_fib_lookup failed (no L2 resolution for the NH)
+func HaveFibIfindex() (err error) {
+	defer func() {
+		if err != nil && !errors.Is(err, ebpf.ErrNotSupported) {
+			log.WithError(err).Fatal("failed to probe HaveFibIfindex")
```

Comment:
I honestly don't know. I am new to golang (literally my first program).

I've seen the same approach being used in other probes (E.g. https://github.com/cilium/cilium/blob/4733a894097f79d98c9872cc85b3ca07c6a6b188/pkg/datapath/linux/probes/probes.go#L669), and I thought it was kind of a good pattern to have a unified end trace of error in case the probe failed.

Not sure if at the end it will be better (e.g. copying the string on each err. path print)

Reply:
No worries! Here is how I think about this:
- The Have* functions are meant to detect whether a feature is present or not. This means that the calling code will adapt to an error being returned. This would bias me towards not logging in the function itself.
- Feature probes are a bit tricky since they can actually have three separate outcomes: supported, not supported, inconclusive (due to some unforeseen error). The code you've linked to only logs the inconclusive case, which makes sense to me.

To sum up: I would remove the logging throughout the function and keep this defer call at the top.

Reply:
P.S. Due to how the `defer` works you are actually logging twice. `defer func() {}()` creates a closure which has access to the __return variables___ of the enclosing function. See this example: https://go.dev/play/p/6tiSrxY2_bu

See the language spec: https://go.dev/ref/spec#Defer_statements
> That is, if the surrounding function returns through an explicit [return statement](https://go.dev/ref/spec#Return_statements), deferred functions are executed after any result parameters are set by that return statement but before the function returns to its caller.

Reply:
>To sum up: I would remove the logging throughout the function and keep this defer call at the top.

Logging on error paths - being golang or not - is nothing new. Error specifics have better context locally, or closer to the error as possible (either through direct logging or returning the error with the string).

I think the _reason why_ a probe has failed is important to be printed as probes in general shall NOT fail. They can evaluate to true or false, but they shall always work. So I very much think somehow the specific error must be printed, somewhere.

I am fine removing the DBG trace and I am also fine moving specific error strings to `err`, so that they can later be dumped in the `defer()`. Deal?

> P.S. Due to how the defer works you are actually logging twice. defer func() {}() creates a closure which has access to the return variables_ of the enclosing function. See this example: https://go.dev/play/p/6tiSrxY2_bu

Yes, I am familiar with `defer()` (other langs have similar constructs), and as I explain above, it was kind of intentional. But I agree it's probably redundant, so I can move the string to `err`.

Reply:
> I think the _reason why_ a probe has failed is important to be printed as probes in general shall NOT fail.

Yes, the canonical way to do that is `fmt.Errorf("more context: %w", err)` or similar, not log statements (imo).

Reply:
Done